### DISCUSSION TITLE
You should render providers as deep as possible in the tree

### DISCUSSION THREAD

**Discussion by Taloqq** [vercel/next.js#71490] [Help]
### Summary

In the docs: https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns

It recommends:  

> You should render providers as deep as possible in the tree – notice how ThemeProvider only wraps {children} instead of the entire <html> document. This makes it easier for Next.js to optimize the static parts of your Server Components.

What does it mean it makes it "**easier**" for Next.js to optimize the static parts of your Server Components?

Easier in what sense? Is it sometimes unable to optimize the static parts because of the context provider? Does it compile slower? 

I'm considering wrapping my root layout with context provider to provide application-wide modals and related events, should I be wary of something?

**Comment by amplicity** [comment]
I don't believe you should wrap your root layout with context provider. this will force a re-render everytime the context changes.

**Comment by icyJoseph** [comment]
Interesting. I also find the language used there, a bit loose.

Perhaps it refers to, one accidental issue, with providers that don't render their children, immediately. For example a Provider that'd do this:

```jsx
// within a Provider
  useEffect(() => { 
   /* something happens that ends with */ 
    setIsLoading(false)
  }, [])

  if (isLoading) {
    return <div>Loading...</div>;
  }
  return children;
```

Then, yeah, during SSR, that children is not gonna be part of the HTML, rather, on the client, once the state is set, within the effect, you'd see the children (page). 

Would be nice to tag the author, Delba, and get some comments from them as well.
